File: src/lib/plugins.ts
import { invoke } from "@tauri-apps/api/core"

export interface Plugin {
  id: string
  name: string
  description: string
  prefix: string
  icon: string
  config?: PluginConfig
  search: (query: string) => Promise<PluginResult[] | PluginHtmlResult>
  onPrefixActivate?: () => void
}

export interface PluginHtmlResult {
  html: string
}

export interface PluginConfig {
  [key: string]: any
}

export interface PluginResult {
  id: string
  title: string
  subtitle?: string
  icon?: string
  actions: PluginAction[]
}

export interface PluginAction {
  id: string
  label: string
  shortcut?: string
}

export async function loadPlugins(): Promise<Plugin[]> {
  try {
    const availablePlugins = await invoke<Plugin[]>("list_plugins")
    const settings = await invoke("get_settings") as any
    const enabledPlugins = settings?.enabledPlugins || {}
    
    return availablePlugins
      .filter(plugin => enabledPlugins[plugin.id] !== false)
      .map(pluginInfo => ({
        ...pluginInfo,
        search: async (query: string) => {
          return await invoke<PluginResult[] | PluginHtmlResult>("search_plugin", { pluginId: pluginInfo.id, query })
        }
      }))
  } catch (error) {
    console.error("Failed to load plugins:", error)
    return []
  }
}

export async function executePluginAction(pluginId: string, resultId: string, actionId: string): Promise<string> {
  return await invoke<string>("execute_plugin_action", { pluginId, resultId, actionId })
}

File: plugins/plugins.rs
use std::collections::HashMap;

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct PluginConfig {
    #[serde(flatten)]
    pub data: HashMap<String, serde_json::Value>,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct PluginAction {
    pub id: String,
    pub label: String,
    pub shortcut: Option<String>,
}

#[derive(serde::Serialize, Clone)]
pub struct PluginResult {
    pub id: String,
    pub title: String,
    pub subtitle: Option<String>,
    pub icon: Option<String>,
    pub actions: Option<Vec<PluginAction>>,
}

#[derive(serde::Serialize, Clone)]
pub struct PluginHtmlResult {
    pub html: String,
}

#[derive(serde::Serialize, Clone)]
#[serde(untagged)]
pub enum PluginSearchResult {
    Results(Vec<PluginResult>),
    Html(PluginHtmlResult),
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub struct Plugin {
    pub id: String,
    pub name: String,
    pub description: String,
    pub prefix: String,
    pub icon: String,
    pub config: Option<PluginConfig>,
}

pub trait PluginTrait: Send + Sync {
    fn get_info(&self) -> Plugin;
    fn search(&self, query: &str) -> Vec<PluginResult>;
    fn execute_action(&self, result_id: &str, action_id: &str) -> Result<String, String>;
}
